File: mobile/flutter/lib/token.dart
import 'package:http/http.dart' as http;
import 'dart:convert';

import 'constants.dart';

class Token {
  String appToken;
  String cardName;
  String clientId;
  String clientSecret;
  Map data;
  Map accessToken;

  Token({this.clientId, this.clientSecret, this.data, this.cardName});

  Future<void> getUserToken() async {
    String credentials = '$clientId:$clientSecret';
    Codec<String, String> stringToBase64 = utf8.fuse(base64);
    String authHeader = stringToBase64.encode(credentials);
    Uri targetURL;
    Uri clientTokenURL;

    if (isClientTokenOAuth) {
      clientTokenURL =
          Uri.parse('https://${url()}/authorization/v1/client-token');
    } else {
      String username = data['username'];
      String password = data['password'];

      var nonAdminURL = Uri.parse(
          'https://${url()}/authorization/v1/oauth/token?grant_type=password&username=$username&password=$password');

      var adminURL = Uri.parse(
          'https://${url()}/authorization/v1/oauth/token?grant_type=client_credentials');

      if (cardName == 'card_admin') {
        targetURL = adminURL;
      } else {
        targetURL = nonAdminURL;
      }
    }

    try {
      String clientToken = data["clientToken"];
      var userTokenData = isClientTokenOAuth
          ? await http.post(clientTokenURL, headers: {
              'Authorization': 'Basic $authHeader',
              'client-token': 'Bearer $clientToken'
            })
          : await http
              .post(targetURL, headers: {'Authorization': 'Basic $authHeader'});

      accessToken = json.decode(userTokenData.body);
      getAppToken(cardName, accessToken['access_token']);
    } on Exception catch (e) {
      print('Something gone wrong with HTTP Request: $e');
    }
  }

  Future<void> getAppToken(cardName, userToken) async {
    try {
      var appTokenUrl =
          Uri.parse('https://${url()}/admin/v1/app_token?app_name=$cardName');

      var appTokenDate = await http
          .get(appTokenUrl, headers: {'Authorization': 'Bearer ${userToken}'});

      Map tokenResponse = (json.decode(appTokenDate.body))[0];
      appToken = tokenResponse['app_token'].toString();
    } on Exception catch (e) {
      print('Something gone wrong with HTTP Request: $e');
    }
  }
}

File: mobile/xamarin/xamarin(.Net)/HydrogenCards/Token.cs
using System;
using System.IO;
using System.Text;
using Newtonsoft.Json;
using System.Collections.Generic;
using System.Linq;
using System.Security.Claims;
using System.Security.Cryptography;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.OpenSsl;
using Org.BouncyCastle.Security;

namespace HydrogenCards
{
    public class AppToken
    {
        [JsonProperty("app_token")]
        public string app_token { get; set; }
    }

    public class AccessToken
    {
        [JsonProperty("access_token")]
        public string access_token { get; set; }
    }

    public class Token
    {
        public string CreateToken(List<Claim> claims, string privateRsaKey)
        {
            RSAParameters rsaParams;
            using (var tr = new StringReader(privateRsaKey))
            {
                var pemReader = new PemReader(tr);

                var keyPair = pemReader.ReadObject() as RsaPrivateCrtKeyParameters;
                if (keyPair == null)
                {
                    throw new Exception("Could not read RSA private key");
                }
                var privateRsaParams = keyPair;
                rsaParams = DotNetUtilities.ToRSAParameters(privateRsaParams);
            }
            using (RSACryptoServiceProvider rsa = new RSACryptoServiceProvider())
            {
                rsa.ImportParameters(rsaParams);
                Dictionary<string, object> payload = claims.ToDictionary(k => k.Type, v => (object)v.Value);
                return Jose.JWT.Encode(payload, rsa, Jose.JwsAlgorithm.RS512);
            }
        }


        public string GetAppToken(string appName, string token)
        {
            string WEBSERVICE_URL = $"https://{Constants.ENV}.hydrogenplatform.com/admin/v1/app_token?app_name={appName}";
            try
            {
                var webRequest = System.Net.WebRequest.Create(WEBSERVICE_URL);
                var appToken = "";
                if (webRequest != null)
                {
                    webRequest.Method = "GET";
                    webRequest.Timeout = 12000;
                    webRequest.ContentType = "application/json";
                    webRequest.Headers.Add("Authorization", "Bearer " + token);

                    using (System.IO.Stream s = webRequest.GetResponse().GetResponseStream())
                    {
                        using (System.IO.StreamReader sr = new System.IO.StreamReader(s))
                        {
                            var jsonResponse = sr.ReadToEnd();
                            var deserialized = JsonConvert.DeserializeObject<AppToken[]>(jsonResponse);
                            appToken = deserialized[0].app_token;
       
                            return appToken;
                        }
                    }
                }
                return appToken;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.ToString());
            }
            return "";
        }



        public string WebRequest(string username, string password)
        {
            string WEBSERVICE_URL = $"https://{Constants.ENV}.hydrogenplatform.com/authorization/v1/oauth/token?grant_type=password&username={username}&password={password}";
            string svcCredentials = Convert.ToBase64String(ASCIIEncoding.ASCII.GetBytes(Constants.client_id + ":" + Constants.client_secret));

            try
            {
                var webRequest = System.Net.WebRequest.Create(WEBSERVICE_URL);
                var userToken = "";

                if (webRequest != null)
                {
                    webRequest.Method = "POST";
                    webRequest.Timeout = 12000;
                    webRequest.ContentType = "application/json";
                    webRequest.Headers.Add("Authorization", "Basic " + svcCredentials);

                    using (Stream s = webRequest.GetResponse().GetResponseStream())
                    {
                        using (StreamReader sr = new StreamReader(s))
                        {
                            var jsonResponse = sr.ReadToEnd();
                            var deserialized = JsonConvert.DeserializeObject<AccessToken>(jsonResponse);

                            userToken = deserialized.access_token;
                            return userToken;
                        }
                    }
                }
                return userToken;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.ToString());
            }
            return "";
        }

        public string AdminWebRequest()
        {
            string WEBSERVICE_URL = $"https://{Constants.ENV}.hydrogenplatform.com/authorization/v1/oauth/token?grant_type=client_credentials";
            string svcCredentials = Convert.ToBase64String(ASCIIEncoding.ASCII.GetBytes(Constants.client_id + ":" + Constants.client_secret));

            try
            {
                var webRequest = System.Net.WebRequest.Create(WEBSERVICE_URL);
                var userToken = "";

                if (webRequest != null)
                {
                    webRequest.Method = "POST";
                    webRequest.Timeout = 12000;
                    webRequest.ContentType = "application/json";
                    webRequest.Headers.Add("Authorization", "Basic " + svcCredentials);

                    using (Stream s = webRequest.GetResponse().GetResponseStream())
                    {
                        using (StreamReader sr = new StreamReader(s))
                        {
                            var jsonResponse = sr.ReadToEnd();
                            var deserialized = JsonConvert.DeserializeObject<AccessToken>(jsonResponse);

                            userToken = deserialized.access_token;
                            return userToken;
                        }
                    }
                }
                return userToken;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.ToString());
            }
            return "";
        }


        public string ClientTokenWebRequest(string clientToken)
        {
            string WEBSERVICE_URL = $"https://{Constants.ENV}.hydrogenplatform.com/authorization/v1/client-token";
            string svcCredentials = Convert.ToBase64String(ASCIIEncoding.ASCII.GetBytes(Constants.client_id + ":" + Constants.client_secret));

            try
            {
                var webRequest = System.Net.WebRequest.Create(WEBSERVICE_URL);
                var userToken = "";

                if (webRequest != null)
                {
                    webRequest.Method = "POST";
                    webRequest.Timeout = 12000;
                    webRequest.ContentType = "application/json";
                    webRequest.Headers.Add("Authorization", "Basic " + svcCredentials);
                    webRequest.Headers.Add("Client-Token", "Bearer " + clientToken);

                    using (Stream s = webRequest.GetResponse().GetResponseStream())
                    {
                        using (StreamReader sr = new StreamReader(s))
                        {
                            var jsonResponse = sr.ReadToEnd();
                            var deserialized = JsonConvert.DeserializeObject<AccessToken>(jsonResponse);

                            userToken = deserialized.access_token;
                            return userToken;
                        }
                    }
                }
                return userToken;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.ToString());
            }
            return "";
        }
    }
}